File: dsl-based/lib/raspp/output.rb
# frozen_string_literal: true
#
# This file is part of Raspp.
# Copyright (C) 2016 Jeffrey Sharp
#
# Raspp is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published
# by the Free Software Foundation, either version 3 of the License,
# or (at your option) any later version.
#
# Raspp is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
# the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Raspp.  If not, see <http://www.gnu.org/licenses/>.
#

class Output
  attr_reader :out, :err

  def initialize(out = nil, err = nil)
    @out = out || $stdout
    @err = err || $stderr
  end

  def warnings?; !!@warnings; end
  def errors?;   !!@errors;   end

  def log_info(*args)
    write_message :'INFO', *args
  end

  def log_warning(*args)
    write_message :WARNING, *args
    @warnings = true
  end

  def log_error(*args)
    write_message :ERROR, *args
    @errors = true
  end

  def write_message(severity, *args)
    loc = source_location
    @out.puts *args.map { |arg| "#{loc}: #{severity}: #{arg}" }
  end

  def write_label(sym)
    raise 'method not implemented'
  end

  def write_instruction(opcode, *args)
    write_directive opcode, *args
  end

  def write_directive(name, *args)
    @out.puts args.empty? \
      ? "\t#{op}"
      : "\t#{op}\t#{args.join(', ')}"
  end

  def source_location(depth = 0)
    loc = ::Kernel.caller_locations(depth + 6, 1).first;
    "#{loc.path}:#{loc.lineno}"
  end
end

##
# GNU Assembler
#
class Gas < Output
  def local_prefix
    '.L.'
  end

  def write_label(sym)
    @out.puts "#{sym}:"
  end

  def define_symbol(name, value)
    write_directive :'.equ', name, value
  end
end

##
# vasm with Motorola syntax
#
class VasmMot < Output
  def local_prefix
    '._'
  end

  def write_label(sym)
    @out.puts sym
  end

  def define_symbol(name, value)
    @out.puts 
    write_directive :'equ', name, value
  end
end

File: rust/src/aspp/output.rs
// This file is part of aspp, a preprocessor for GNU Assembler.
// Copyright (C) 2016 Jeffrey Sharp
//
// aspp is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published
// by the Free Software Foundation, either version 3 of the License,
// or (at your option) any later version.
//
// aspp is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
// the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with aspp.  If not, see <http://www.gnu.org/licenses/>.

use std::fmt::{self, Write};

pub struct Output {
    text: String,
    name: String,
    line: usize,
    bol:  bool,
}

impl Output {
    pub fn with_capacity(name: String, cap: usize) -> Self {
        Output {
            text: String::with_capacity(cap),
            name: name,
            line: 1,
            bol:  true,
        }
    }

    #[inline]
    pub fn as_str(&self) -> &str {
        self.text.as_str()
    }

    #[inline]
    pub fn advance_line(&mut self) {
        self.line += 1
    }

    #[inline]
    pub fn write_raw_char(&mut self, c: char) {
        self.text.push(c)
    }

    #[inline]
    pub fn write_raw_str(&mut self, s: &str) {
        self.text.push_str(s)
    }

    pub fn write_local_symbol(&mut self, name: &str) {
        write!(self.text, "L({})", name).unwrap()
    }

    pub fn write_local_label(&mut self, name: &str) {
        write!(self.text, "L({}):", name).unwrap()
    }

    pub fn label(&mut self, name: &str) {
        write!(self.text,
               "\n\
                ; SCOPE: {2}\n\
                #ifdef scope\n\
                #undef scope\n\
                #endif\n\
                #define scope {0}\n\
                \n\
                # {0} \"{1}\"\n\
                .label scope;",
               self.line, self.name, name
              ).unwrap()
    }

    pub fn global(&mut self) {
        write!(self.text,
               "\n\
                # {0} \"{1}\"\n\
                .global scope;",
               self.line, self.name
              ).unwrap()
    }
}

impl Write for Output {
    #[inline(always)]
    fn write_str(&mut self, s: &str) -> Result<(), fmt::Error> {
        self.text.write_str(s)
    }

    #[inline(always)]
    fn write_char(&mut self, c: char) -> Result<(), fmt::Error> {
        self.text.write_char(c)
    }

    #[inline(always)]
    fn write_fmt(&mut self, args: fmt::Arguments) -> Result<(), fmt::Error> {
        self.text.write_fmt(args)
    }
}

